pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,274 @@
+  /// session. For example: undo history.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Caches/TemporaryItems/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_RUNTIME_HOME/<tool>` if `$XDG_RUNTIME_HOME` is defined, and
+  ///   - `$HOME/.cache/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap [cacheHome] and [stateHome] or be
+  /// nested in [cacheHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String runtimeHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.runtime), tool);
+
+  /// The path of the directory where [tool] can place its state.
+  ///
+  /// The state directory is likely not backed up or synchronized accross
+  /// devices by the OS. For data that may be backed up and synchronized, use
+  /// [dataHome] instead.
+  ///
+  /// This is a location appropriate for storing data which is either not
+  /// important enougn, not small enough, or not portable enough to store in
+  /// [dataHome]. For example: logs and indices.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Application Support/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_STATE_HOME/<tool>` if `$XDG_STATE_HOME` is defined, and
+  ///   - `$HOME/.local/state/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [cacheHome], and
+  /// [runtimeHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String stateHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.state), tool);
+
+  String _baseDirectory(_XdgBaseDirectoryType dir) {
+    if (Platform.isWindows) {
+      return _baseDirectoryWindows(dir);
+    }
+    if (Platform.isMacOS) {
+      return _baseDirectoryMacOs(dir);
+    }
+    if (Platform.isLinux) {
+      return _baseDirectoryLinux(dir);
+    }
+    throw UnsupportedError('Unsupported platform: ${Platform.operatingSystem}');
+  }
+
+  String _baseDirectoryWindows(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config ||
+        _XdgBaseDirectoryType.data =>
+          _requireEnv('APPDATA'),
+        _XdgBaseDirectoryType.cache ||
+        _XdgBaseDirectoryType.runtime ||
+        _XdgBaseDirectoryType.state =>
+          _requireEnv('LOCALAPPDATA'),
+      };
+
+  String _baseDirectoryMacOs(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config =>
```

Comment:
MacOS uses `/` as separator, so a single `'Library/Preferences'` should work.
Then it can be stored on the enum value and this function can become `=> path.join(_home, dir._macOSDefaultPath);`.